# **Human Activity Recognition**

## ***Problem statement :***
### John wants to find out the amount of calories he is burning everyday, for this, he needs to know the various activities he’s doing all day using the means of his smartphone. Inorder to do this he collects the Accelerometer and Gyroscope readings built into his smartphoneand applies several machine learning algorithms to find out the best algorithm to predict and classify his daily activities.

In [1]:
from google.colab import drive 
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [6]:
# importing required libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [7]:
train_data = pd.read_csv('/content/drive/My Drive/Projects/Human activity detection/UCI_HAR_Dataset/csv_files/train.csv')
test_data = pd.read_csv('/content/drive/My Drive/Projects/Human activity detection/UCI_HAR_Dataset/csv_files/test.csv')
print('file read')

file read


In [8]:
train_data.head()

tBodyAccmeanX  tBodyAccmeanY  tBodyAccmeanZ  ...  subject  Activity  ActivityName
0       0.288585      -0.020294      -0.132905  ...        1         5      STANDING
1       0.278419      -0.016411      -0.123520  ...        1         5      STANDING
2       0.279653      -0.019467      -0.113462  ...        1         5      STANDING
3       0.279174      -0.026201      -0.123283  ...        1         5      STANDING
4       0.276629      -0.016570      -0.115362  ...        1         5      STANDING

[5 rows x 564 columns]

In [9]:
train_data.Activity.value_counts()

6    1407
5    1374
4    1286
1    1226
2    1073
3     986
Name: Activity, dtype: int64

In [10]:
train_data.shape

(7352, 564)

In [11]:
# shuffling the data for predictions
from sklearn.utils import shuffle
train_data = shuffle(train_data)

In [12]:
# creating labels and outputs
X_train = train_data.drop(['subject', 'Activity', 'ActivityName'], axis = 1)
y_train = train_data.ActivityName

X_test = test_data.drop(['subject', 'Activity', 'ActivityName'], axis = 1)
y_test = test_data.ActivityName

print('Successfully created labels and outputs')


Successfully created labels and outputs


### Now finding the best model for this data by checking the accuracy score for each and then picking the one with the highest score.

## *1. LOGISTIC REGRESSION MODEL*

In [13]:
# Logistic regression model
from sklearn.linear_model import LogisticRegression
lreg = LogisticRegression()
lreg.fit(X_train, y_train)

/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [15]:
result = pd.DataFrame({'Actual':y_test, 'Predict':y_pred})
result

Actual           Predict
0             STANDING          STANDING
1             STANDING          STANDING
2             STANDING          STANDING
3             STANDING          STANDING
4             STANDING          STANDING
...                ...               ...
2942  WALKING_UPSTAIRS  WALKING_UPSTAIRS
2943  WALKING_UPSTAIRS  WALKING_UPSTAIRS
2944  WALKING_UPSTAIRS  WALKING_UPSTAIRS
2945  WALKING_UPSTAIRS  WALKING_UPSTAIRS
2946  WALKING_UPSTAIRS  WALKING_UPSTAIRS

[2947 rows x 2 columns]

In [41]:
# accuracy check
from sklearn.metrics import accuracy_score
pred1 = lreg.predict(X_test)
print('Accuracy for Logistic regression : ', accuracy_score(y_test,pred1))

Accuracy for Logistic regression :  0.9592806243637597


**Accuracy for Logistic regression :  0.9592806243637597**

## *2. DECISION TREE CLASSIFIER*

In [30]:
# Decision tree classifier
from sklearn.tree import DecisionTreeClassifier
dtmodel = DecisionTreeClassifier()
dtmodel.fit(X_train, y_train)


DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=None, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=None, splitter='best')

In [31]:
# accuracy check
from sklearn.metrics import accuracy_score
pred2 = dtmodel.predict(X_test)
print('Accuracy for Decision Tree : ',accuracy_score(y_test,pred2))

Accuracy for Decision Tree :  0.8605361384458772


**Accuracy for Decision Tree :  0.8605361384458772**



## *3. K-NEAREST NEIGHBOURS*

In [32]:
# K-nearest neighbours
from sklearn.neighbors import KNeighborsClassifier
kmodel = KNeighborsClassifier()
kmodel.fit(X_train, y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                     metric_params=None, n_jobs=None, n_neighbors=5, p=2,
                     weights='uniform')

In [33]:
# accuracy check 
from sklearn.metrics import accuracy_score
pred3 = kmodel.predict(X_test)
print('Accuracy for KNN : ',accuracy_score(y_test,pred3))

Accuracy for KNN :  0.9002375296912114


**Accuracy for KNN :  0.9002375296912114**

## *4. NAIVE BAYES MODEL*

In [35]:
# Naive bayes 
from sklearn.naive_bayes import GaussianNB
nbmodel = GaussianNB()
nbmodel.fit(X_train, y_train)

GaussianNB(priors=None, var_smoothing=1e-09)

In [37]:
# accuracy check
from sklearn.metrics import accuracy_score
pred4 = nbmodel.predict(X_test)
print('Accuracy for Naive bayes : ', accuracy_score(y_test,pred4))

Accuracy for Naive bayes :  0.7702748557855447


**Accuracy for Naive bayes :  0.7702748557855447**

## *5. Support Vector Machine*

In [39]:
# Suport vector machine
from sklearn import svm
svmodel = svm.SVC(kernel = 'linear')
svmodel.fit(X_train, y_train)

SVC(C=1.0, break_ties=False, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma='scale', kernel='linear',
    max_iter=-1, probability=False, random_state=None, shrinking=True,
    tol=0.001, verbose=False)

In [40]:
# Accuracy check
from sklearn.metrics import accuracy_score 
pred5 = svmodel.predict(X_test)
print('Accuracy for SVM model : ', accuracy_score(y_test,pred5))

Accuracy for SVM model :  0.9640312181879878


**Accuracy for SVM model :  0.9640312181879878**

### Out of all the classification models the best accuracy score was given by the Support Vestor Machine model for the given dataset.
### So the SVM is the best algorithm to predict and classify the daily activities from the dataset